python/dgl/transform/module.py

```diff
@@ -0,0 +1,790 @@
+        """
+        if self.combine_like:
+            src, dst = g.edges(etype=c_etype)
+            src, dst = F.cat([src, dst], dim=0), F.cat([dst, src], dim=0)
+            data_dict[c_etype] = (src, dst)
+        else:
+            self.transform_asymmetric_etype(c_etype, g, data_dict)
+
+    def transform_asymmetric_etype(self, c_etype, g, data_dict):
+        r"""
+
+        Description
+        -----------
+        Transform the graph corresponding to an asymmetric canonical edge type.
+
+        Parameters
+        ----------
+        c_etype : tuple of str
+            A canonical edge type.
+        g : DGLGraph
+            The graph.
+        data_dict : dict
+            The edge data to update.
+        """
+        utype, etype, vtype = c_etype
+        src, dst = g.edges(etype=c_etype)
+        data_dict.update({
+            c_etype: (src, dst),
+            (vtype, 'rev_{}'.format(etype), utype): (dst, src)
+        })
+
+    def transform_etype(self, c_etype, g, data_dict):
+        r"""
+
+        Description
+        -----------
+        Transform the graph corresponding to a canonical edge type.
+
+        Parameters
+        ----------
+        c_etype : tuple of str
+            A canonical edge type.
+        g : DGLGraph
+            The graph.
+        data_dict : dict
+            The edge data to update.
+        """
+        utype, _, vtype = c_etype
+        if utype == vtype:
+            self.transform_symmetric_etype(c_etype, g, data_dict)
+        else:
+            self.transform_asymmetric_etype(c_etype, g, data_dict)
+
+    def __call__(self, g):
+        data_dict = dict()
+        for c_etype in g.canonical_etypes:
+            self.transform_etype(c_etype, g, data_dict)
+        new_g = update_graph_structure(g, data_dict, copy_edata=False)
+
+        # Copy and expand edata
+        for c_etype in g.canonical_etypes:
+            utype, etype, vtype = c_etype
+            if utype == vtype and self.combine_like:
+                for key, feat in g.edges[c_etype].data.items():
+                    new_feat = feat if self.copy_edata else F.zeros(
+                        F.shape(feat), F.dtype(feat), F.context(feat))
+                    new_g.edges[c_etype].data[key] = F.cat([feat, new_feat], dim=0)
+            else:
+                rev_c_etype = (vtype, 'rev_{}'.format(etype), utype)
+                for key, feat in g.edges[c_etype].data.items():
+                    new_g.edges[c_etype].data[key] = feat
+                    if self.copy_edata:
+                        new_g.edges[rev_c_etype].data[key] = feat
+
+        return new_g
+
+class ToSimple(BaseTransform):
+    r"""
+
+    Description
+    -----------
+    Convert a graph to a simple graph without parallel edges and return a new graph.
+
+    Parameters
+    ----------
+    return_counts : str, optional
+        The edge feature name to hold the edge count in the original graph.
+    writeback_mapping : bool, optional
```

Comment:
This shouldn't be allowed right? It cannot be used together with `Compose`.

Reply:
Removed.